[Koh Rui Ling] iP

ruilingk opened the pull request:
Pull request for week 3 iP task.

Commit:
Add Gradle support

ruilingk committed:
Add Level-1

ruilingk committed:
Add Level-2 and personalised the bot

ruilingk committed:
Add Task.java

ruilingk committed:
Update Duke.java

Extracted out the codes for different commands (e.g. "list", "done", "bye")
into separate methods within Duke class.

ruilingk committed:
Add Level-3

ruilingk committed:
Add ToDo.java

Modified Duke class to incorporate adding of ToDo task.
Removed toString() method in Task class.

ruilingk committed:
Add Deadline.java

Modified Duke class to incorporate adding of Deadline task.
Added toString() method back into Task class.

ruilingk committed:
Add Level-4

Added Event class.
Modified Duke class to incorporate adding of Event task.

ruilingk committed:
Add Automated Text UI Testing

Modified runtest.bat
Modified EXPECTED.TXT and input.txt for testing purposes.

ruilingk committed:
Add DukeErrorsHandler.java

Modified Duke.java to handle errors.

ruilingk committed:
Update input.txt and EXPECTED.TXT

ruilingk committed:
Add Level-5

Added DukeCommandsHandler class to handle commands.
Modified DukeException class.

ruilingk committed:
Add Level-6

Updated DukeCommandsHandler class to incorporate deletion of tasks.
Updated DukeException class with clearer error messages.
Updated input.txt and EXPECTED.TXT

ruilingk committed:
Personalise bot

Added more personality to the bot.
Updated EXPECTED.TXT and input.txt

ruilingk committed:
Replace tick and cross symbols with Y and N respectively

ruilingk committed:
Replace tick and cross symbols with ✓ and ✘ respectively

ruilingk committed:
Add Storage.java

Bot now can save and load data from user's directory.

ruilingk committed:
Add Level-7

Updated DukeExceptions class so that exceptions are thrown now.

ruilingk committed:
Add Level-8

Add date and time functionality.
Deadline and Event objects now have associated date and time stamps, so
that users can input and bot can sense it.
Handled exceptions for when users input incorrect format of date and
time.

ruilingk committed:
Merge branch 'branch-Level-7' into master

ruilingk committed:
Merge branch 'branch-Level-8' into master

ruilingk committed:
Add Level-8 after merging

Updated input and EXPECTED text files to reflect most recent changes.

ruilingk committed:
Add TaskList.java

TaskList class has properties to add tasks, delete tasks, mark tasks
done, and listing tasks.
Made minor updates to other classes to incorporate the changes.

ruilingk committed:
Add UI.java

UI class has the methods that is in charge of interacting with the user
for example, greeting of user and exiting.
Made minor updates to other classes to incorporate the changes.

ruilingk committed:
Add Parser.java and Command.java

Added all the other relevant commands classes too, to better organise
the code since all commands have their own behaviours.

ruilingk committed:
Add A-MoreOOP

Moved textToTask method from Task class to Parser class.
Replaced methods with public access modifier to be protected.

ruilingk committed:
Add A-Packages

ruilingk committed:
Add JUnit tests for all commands

ruilingk committed:
Add A-JUnit

ruilingk committed:
Add A-JavaDoc

ruilingk committed:
Add A-CodingStandard

ruilingk committed:
Add Level-9

ruilingk committed:
Add Level-9

ruilingk committed:
Update Parser.java

ruilingk committed:
Merge branch 'branch-A-JavaDoc' into master

ruilingk committed:
Merge branch 'branch-A-CodingStandard' into master

ruilingk committed:
Add A-CodingStandard after merging

ruilingk committed:
Add Level-9 after merging

ruilingk committed:
Add HelpCommand.java

HelpCommand class can help first-time users to understand the different
commands that Focus can do, so it will be easier for them.
Made modifications to some classes to incorporate the changes, and
updated JUnit tests, EXPECTED.txt and input.txt

ruilingk committed:
Remove comments from Duke.java

ruilingk committed:
Make minor adjustment to UI.java

ruilingk committed:
Merge remote-tracking branch 'origin/add-gradle-support' into master

ruilingk committed:
Add A-Gradle

Added Gradle and Checkstyle plugin.
Gradle can now run, build and run unit tests.

ruilingk committed:
Add A-CheckStyle

ruilingk committed:
Add skeleton structure for GUI

Followed the JavaFX guide.

ruilingk committed:
Add functionality for different commands in GUI

Greetings to user is not yet implemented.
Graphical interface is not as nice yet.

ruilingk committed:
Add greetings functionality to GUI

Changed fonts and font size, and size of panel. Images are also in
circle formats.

ruilingk committed:
Merge branch 'branch-A-Gradle' into master

ruilingk committed:
Merge branch 'branch-A-CheckStyle' into master

ruilingk committed:
Merge branch 'branch-Level-10' into master

ruilingk committed:
Add A-CheckStyle after merging

ruilingk committed:
Add new line after description of JavaDocs

UnicornJin commented on src/main/java/DialogBox.java:
```diff
@@ -0,0 +1,86 @@
+    }
+
+    /**
+     * Creates a new DialogBox for user.
```
Good JavaDoc

UnicornJin commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,58 @@
+ * Main class to run Focus.
+ */
 public class Duke {
-    public static void main(String[] args) {
```
Haha you made a logo for Duke, interesting!

UnicornJin commented on src/main/java/Main.java:
```diff
@@ -0,0 +1,43 @@
+     */
+    private final Duke duke = new Duke();
+    /**
+     * Gets the image for Duke.
```
you can just use /* */ for one line JavaDoc

ruilingk replied:
Okay thank you for pointing it out!

UnicornJin commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,41 @@
+package duke.command;
```
Good, using different packages for different usage of classes

ruilingk replied:
Thanks!

UnicornJin reviewed (commented):
Good code, hope your program runs well!

ureshiiYing commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,29 @@
+/**
+ * Represents all the commands to allow different command classes to inherit.
+ */
+public abstract class Command {
```
I like that how you abstracted out the commands into a Command class, and made it an abstract class.

ruilingk replied:
Thanks!

ureshiiYing commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,77 @@
+    /**
+     * '0' to represent incomplete, '1' to represent completed.
+     */
+    protected String completed;
```
Perhaps this String variable is not necessary as you already have a boolean to represent whether the task is completed.

ruilingk replied:
Hi, I had this extra string variable because it is used to convert the boolean to string which is to be placed into the text file when user starts running the app. Hope this clarifies!

ureshiiYing commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,77 @@
+    /**
+     * Description of task.
+     */
+    protected String taskName;
```
Perhaps a more intuitive variable name could be used, eg. taskDescription? It will match your javadoc comment too.

ruilingk replied:
Alright thank you!

ureshiiYing commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,77 @@
+     * @return Empty string.
+     */
+    public String taskToText() {
+        return "";
```
I got confused here, why do you return an empty string if this method is convert the task to a string form that is to be saved?
Update: I see you that you've overridden this method in the child classes. I suggests turning the Task class into an abstract class since a Task is not instantiated, while its child classes are.

ureshiiYing commented on src/main/java/duke/ui/UI.java:
```diff
@@ -0,0 +1,69 @@
+                + "\n\tTo exit Focus:"
+                + "\n\t\t- bye\n";
+        return commands;
+    }
```
Wow! I think this is a really nifty customization that you have added to the iP.

ureshiiYing commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,58 @@
+/**
+ * Main class to run Focus.
+ */
 public class Duke {
```
Perhaps you could've renamed the Duke class to Focus class since you're bot is named Focus Pocus? It is a cute name, by the way.

ruilingk committed:
Change single-line Javadocs to one-liner

ureshiiYing commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,199 @@
+     * @return String representation of added Deadline.
+     * @throws DukeException If input does not meet criteria.
+     */
+    private String addDeadline(String input, TaskList taskList, Storage storage) throws DukeException {
```
I realised that you only have one DukeException class that covers all types of Exception that could be thrown. I feel that these Exceptions could be abstracted out, eg. InappropriateDescriptionException, InappropriateCommandException etc, to be more OOP.

ruilingk replied:
Okay thank you for your suggestion!

ureshiiYing commented on src/main/java/duke/command/InvalidCommand.java:
```diff
@@ -0,0 +1,33 @@
+/**
+ * Represents the InvalidCommand to deal with invalid inputs by user.
+ */
+public class InvalidCommand extends Command {
```
Perhaps this should be an Exception as the Command class should represent the commands that the user can input?

ruilingk replied:
Alright thank you.

ureshiiYing commented on src/main/java/duke/command/ExitCommand.java:
```diff
@@ -0,0 +1,30 @@
+ */
+public class ExitCommand extends Command {
+    /**
+     * Returns true since ExitCommand is not an ExitCommand.
```
I think you had a typo here. It should be is an ExitCommand.

ruilingk replied:
Oh right, thank you!

ureshiiYing reviewed (commented):
Overall, I would like to applaud the effort that you have put into do unit testing for all of the classes! Additionally, your customisations done are cute. I also did not find any coding standard violations, which is good! One thing you could work on is to abstract out the various types of Exceptions in your bot.

ruilingk committed:
Beautify GUI for Focus

ruilingk committed:
Package new classes

ruilingk committed:
Customise Focus further

Dialog boxes now have different colours for Pocus and User.
Images of Pocus and User are now png so it is nicer.
Naming of dialog boxes and tabbing of content so it is easier for
users to differentiate and see.
Window will close automatically after 5 seconds when user exits.

Licheng-Wu commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,50 @@
+     * @return Focus' response.
+     */
+    public String getResponse(String input) {
+        boolean exit;
```
Perhaps having a more easily understandable variable name to make it sound more like a boolean?

ruilingk replied:
Alright noted, I have changed it to `isExit`. Thank you!

Licheng-Wu commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,88 @@
+     */
+    public static Command parse(String input) {
+        Command command;
+        if (input.startsWith("help")) { // list out the commands
+            command = new HelpCommand();
+        } else if (input.startsWith("todo")) { // add to-do tasks
+            command = new AddCommand("todo");
+        } else if (input.startsWith("deadline")) { // add deadline tasks
+            command = new AddCommand("deadline");
+        } else if (input.startsWith("event")) { // add event tasks
+            command = new AddCommand("event");
+        } else if (input.startsWith("delete")) { // delete tasks
+            command = new DeleteCommand();
+        } else if (input.startsWith("done")) { // mark tasks done
+            command = new DoneCommand();
+        } else if (input.equals("list")) { // list out the tasks
+            command = new ListCommand();
+        } else if (input.startsWith("find")) { // find the tasks
+            command = new FindCommand();
+        } else if (input.equals("bye")) { // exit the bot
+            return new ExitCommand();
+        } else { // handle invalid inputs
+            command = new InvalidCommand();
```
Consider using switch statements for better multi-way branching

Licheng-Wu commented on src/test/java/duke/command/DeleteCommandTest.java:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    @Test
+    public void testExecute() {
```
Perhaps using a more intuitive variable name for tests,in the following format: featureUnderTest_testScenario_expectedBehavior()

Licheng-Wu reviewed (commented):
Overall, the code looks good aside from some minor coding standard violation. There is good use of OOP and code is clean and easy to follow.

ruilingk committed:
Create new JAR file using Gradle

ruilingk committed:
Change package and class names to match Focus

Edited some codes in response to peer reviews.

ruilingk committed:
Add A-Assertions

Included assertions for methods that have conditions which should hold
true at certain points in time.

ruilingk committed:
Add different types of Exceptions for Focus

Added different types of exceptions to extend from
FocusException class.
Added Javadocs for the classes.
Updated tests.

ruilingk committed:
Add A-CodeQuality

Refactored code to make it less cluttered, abiding the code quality.
Abstracted out magic literals and extract out methods.

ruilingk committed:
Merge pull request #1 from ruilingk/branch-A-Assertions

Add A-Assertions

ruilingk committed:
Merge branch 'master' into branch-A-CodeQuality

ruilingk committed:
Merge pull request #2 from ruilingk/branch-A-CodeQuality

Add A-CodeQuality

ruilingk committed:
Add A-CodeQuality after merging PR

Made some minor changes to incorporate assertions on new methods.

ruilingk committed:
Add A-CI

ruilingk committed:
Update gradle.yml

ruilingk committed:
Add B-Reminders

Reminders now work if user types in the Remind command.

ruilingk committed:
Add Settings Command

Settings command allows user to change the number of days they want
their reminders to be based on.
Included new JUnit tests and Exception.
Slight modifications to some classes to reflect most recent changes.

ruilingk committed:
Merge branch 'branch-B-Reminders' into master

ruilingk committed:
Add different coloured messages for error messages

ruilingk committed:
Add more personality to Focus

ruilingk committed:
Make minor changes

ruilingk committed:
Add skeleton for User Guide

ruilingk committed:
Update Ui.png and User Guide

ruilingk committed:
Update User Guide with proper images

ruilingk committed:
Set theme jekyll-theme-slate

ruilingk committed:
Set theme jekyll-theme-tactile

ruilingk committed:
Update _config.yml

ruilingk committed:
Update _config.yml

ruilingk committed:
Add new Settings command and update User Guide

Focus now recognises users, and not prompting for their names whenever
they start the application.
Settings command can now change user's name.
User Guide includes content page and acknowledgements.

ruilingk committed:
Merge branch 'master' of https://github.com/ruilingk/ip into master

ruilingk committed:
Modify User Guide

ruilingk committed:
Update README and User Guide

ruilingk committed:
Update Ui.png

ruilingk committed:
Update User Guide with link to Release

ruilingk committed:
Update User Guide

ruilingk committed:
Update image

ruilingk committed:
Update User Guide - Help image
